File: src/main/java/com/br/palpiteiro/service/PalpiteService.java
package com.br.palpiteiro.service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import javax.servlet.http.HttpSession;

import org.springframework.beans.factory.ObjectFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import com.br.palpiteiro.domain.Confronto;
import com.br.palpiteiro.domain.Palpite;
import com.br.palpiteiro.domain.Usuario;
import com.br.palpiteiro.repository.PalpiteRepository;
import com.br.palpiteiro.service.exception.ObjectNotFoundException;

@Service
public class PalpiteService {
	
	@Autowired
	private PalpiteRepository palpiteRepository;
	
	@Autowired
	ObjectFactory<HttpSession> httpSessionFactory;
	
	public Palpite findById(Integer id) {
		
		Optional<Palpite> palpite = palpiteRepository.findById(id);

		return palpite.orElseThrow(() -> new ObjectNotFoundException("Objeto não encontrado! Id: " + id + ", Tipo: " + Palpite.class.getName()));
	}
	
	public List<Palpite> findAll() {
		
		List<Palpite> palpiteList = palpiteRepository.findAll();

		return palpiteList;
	}

	public Palpite save(Palpite palpite) {

		return palpiteRepository.save(palpite);
	}
	
	public List<Palpite> saveAll(List<Confronto> confrontoList) {

		Usuario usuarioLogado = (Usuario) SecurityContextHolder.getContext().getAuthentication().getPrincipal();

		List<Palpite> palpiteList = new ArrayList<Palpite>();
		
		for (Confronto confronto : confrontoList) {
			for (Palpite palpite : confronto.getPalpiteList()) {
				palpite.setConfronto(confronto);
				palpite.setDataHora(LocalDateTime.now());
				palpite.setUsuario(usuarioLogado);
			}

			palpiteList.addAll(confronto.getPalpiteList());
		}
		
		return palpiteRepository.saveAll(palpiteList);
	}

	public void delete(Integer id) {

		palpiteRepository.deleteById(id);
	}

	public Palpite update(Palpite palpite) {

		return save(palpite);
	}
}

File: src/main/frontend/app-angular/src/app/services/palpite.service.ts
import { Injectable } from '@angular/core';

import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Palpite } from '../models/palpite';
import { Confronto } from '../models/confronto';

@Injectable({
  providedIn: 'root'
})

export class PalpiteService {

  private readonly API = 'api/palpites';

  constructor(private httpClient: HttpClient) { }

  findAll(): Observable<Palpite[]> {
	return this.httpClient.get<Palpite[]>(this.API);	
  }
  
  findById(id: number): Observable<Palpite> {
	const url = `${this.API}/${id}`;
	return this.httpClient.get<Palpite>(url);	
  }
  
  save(palpite: Palpite): Observable<Palpite> {
	return this.httpClient.post<Palpite>(this.API, palpite);	
  }
  
  saveAll(confrontoList: Confronto[]): Observable<Palpite[]> {
	return this.httpClient.post<Palpite[]>(this.API, confrontoList);	
  }
  
  update(palpite: Palpite): Observable<Palpite> {
	return this.httpClient.put<Palpite>(this.API, palpite);	
  }
  
  delete(id: number): Observable<void> {
	const url = `${this.API}/${id}`;
	return this.httpClient.delete<void>(url);	
  }
}
